File: ios/ExpoCryptoArgon2Module.swift
import ExpoModulesCore

// Typed record — Expo deserialises the JS object into this struct automatically.
struct Argon2Params: Record {
    @Field var password: String = ""
    @Field var salt: String = ""
    @Field var iterations: Int = 3
    @Field var memory: Int = 65536
    @Field var parallelism: Int = 1
    @Field var hashLength: Int = 32
    @Field var outputType: String = "binary"   // "binary" | "encoded"
}

public class ExpoCryptoArgon2Module: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoCryptoArgon2")

        // AsyncFunction with an explicit Promise so we can dispatch to a global
        // concurrent queue — Argon2 is CPU-heavy and must not block the JS thread.
        AsyncFunction("argon2id") { (params: Argon2Params, promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                // Swift bridges ObjC `NSError **` as a throwing function — use try/catch.
                do {
                    let hex = try Argon2Wrapper.argon2id(
                        withPassword: params.password,
                        salt: params.salt,
                        iterations: params.iterations,
                        memory: params.memory,
                        parallelism: params.parallelism,
                        hashLength: params.hashLength,
                        outputType: params.outputType
                    )
                    promise.resolve(hex)
                } catch {
                    promise.reject(error)
                }
            }
        }
    }
}

File: android/src/main/java/expo/modules/cryptoargon2/ExpoCryptoArgon2Module.kt
package expo.modules.cryptoargon2

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class Argon2Params : Record {
    @Field val password: String = ""
    @Field val salt: String = ""
    @Field val iterations: Int = 3
    @Field val memory: Int = 65536
    @Field val parallelism: Int = 1
    @Field val hashLength: Int = 32
    @Field val outputType: String = "binary"  // "binary" | "encoded"
}

class ExpoCryptoArgon2Module : Module() {

    companion object {
        init {
            System.loadLibrary("expocryptoargon2")
        }
    }

    // Implemented in android/src/main/cpp/argon2-jni.cpp
    private external fun nativeArgon2id(
        password: String,
        salt: String,
        iterations: Int,
        memory: Int,
        parallelism: Int,
        hashLength: Int,
        outputType: String
    ): String

    override fun definition() = ModuleDefinition {
        Name("ExpoCryptoArgon2")

        // AsyncFunction runs on expo-modules-core's background executor.
        // Returning a value resolves the JS Promise; throwing rejects it.
        AsyncFunction("argon2id") { params: Argon2Params ->
            nativeArgon2id(
                params.password,
                params.salt,
                params.iterations,
                params.memory,
                params.parallelism,
                params.hashLength,
                params.outputType
            )
        }
    }
}
